yacl/math/mpint/tommath_ext_features.cc

```diff
@@ -216,16 +213,19 @@ void mp_ext_safe_prime_rand(mp_int *p, int t, int psize) {
     if (!res) {
       continue;
     }
+    MPINT_ENFORCE_OK(mp_prime_is_prime(p, 1, &res));
+    if (!res) {
+      continue;
+    }
     // test Pocklington Criterion
-    if (!is_pocklington_criterion_satisfied(&q)) {
+    if (!is_pocklington_criterion_satisfied(p)) {
       continue;
     }
     // final check
     MPINT_ENFORCE_OK(mp_prime_is_prime(&q, t, &res));
     if (!res) {
       continue;
     }
-    MPINT_ENFORCE_OK(mp_prime_is_prime(p, t, &res));
```

Comment:
@zheyang0825 可能得跟原始 paper 再对一下，虽然 Pocklington Criterion 本身是确定性的，但前提条件 q 是不是 prime 是概率性的，所以最后要不要 mp_prime_is_prime(p) 不确定，如果 paper 没有明确提到，去掉是 ok 的

Reply:
关于mp_prime_is_prime(p, t, &res)我做过一些实验的，包括改变语句的位置、t的大小。如果删去

MPINT_ENFORCE_OK(mp_prime_is_prime(p, 1, &res));

那么前面

// is `q` a prime? try 10 times.
    MPINT_ENFORCE_OK(mp_prime_is_prime(&q, 10, &res));

即使参数t增加到500，即mp_prime_is_prime(&q, 500, &res)，也无济于事。这种奇怪现象我怀疑是libtommath的随机源有问题

Reply:
我仔细看了一下代码，问题不在随机数上面。。而是逻辑有个 bug。。

在 do..while 循环中，如果遇到 continue 语句，C++ 会重新判断一下 while 后的条件，为 true 则执行下一个 loop。

然后问题就出在 222 行的 continue：
```
    if (!is_pocklington_criterion_satisfied(p)) {
      continue;
    }
```

遇到 continue 之后，C++ 执行 `while (!res)`，这个时候 res 变量是陈旧的，上面的语句遗留的，res 值为 true 导致 while 循环退出，continue 没有真正执行。

我认为按以下的方法修改就行：

```
    // is `q` a prime? try 10 times.
    MPINT_ENFORCE_OK(mp_prime_is_prime(&q, 10, &res));
    if (!res) {
      continue;
    }
    // test Pocklington Criterion
    if (!is_pocklington_criterion_satisfied(p)) {
      continue;
    }
    // final check, if q is prime, then p is 100% prime since Pocklington is deterministic
    MPINT_ENFORCE_OK(mp_prime_is_prime(&q, t, &res));
    if (res) {
      return;
    }
  } while (true);
```

Reply:
你的观察非常到位，这样修改确实有效。我在本地测试运行MPIntTest.RandPrimeOverWorks时间是1秒到17秒之间，运气好就是1秒，运气不好就17秒。